### PR TITLE
[PropertyAccess] Allow to disable magic __get & __set

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no     
| Deprecations? | yes 
| Tests pass?   | yes   
| Fixed tickets | N/A   
| License       | MIT
| Doc PR        | TODO

Working with some legacy code having annoying `__get` & `__set` methods, this would have been useful the same way `__call` can be enabled/disabled.

### CODE REVIEW

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 1]
```diff
@@ -71,10 +82,23 @@ class PropertyAccessor implements PropertyAccessorInterface
...
+     *
+     * @param int $magicMethods A bitwise combination of the MAGIC_* constants
+     *                          to specify the allowed magic methods (__get, __set, __call)
+     *                          or self::DISALLOW_MAGIC_METHODS for none
      */
-    public function __construct(bool $magicCall = false, bool $throwExceptionOnInvalidIndex = false, CacheItemPoolInterface $cacheItemPool = null, bool $throwExceptionOnInvalidPropertyPath = true)
+    public function __construct(/*int */$magicMethods = self::MAGIC_GET | self::MAGIC_SET, bool $throwExceptionOnInvalidIndex = false, CacheItemPoolInterface $cacheItemPool = null, bool $throwExceptionOnInvalidPropertyPath = true)
     {
-        $this->magicCall = $magicCall;
+        if (\is_bool($magicMethods)) {
```
Comment (xabbuh) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: Wouldn't it make sense to turn the `$throwExceptionOnInvalidIndex` and `$throwExceptionOnInvalidPropertyPath` into flags too?

Comment (Simperfit) [reply src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: I still think we should see https://github.com/symfony/symfony/issues/31126#issuecomment-484362754

Comment (ogizanagi) [reply src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: I agree as well about this, but I'd rather provide a dedicated PR. Anyway we have the issue to not loose track of it. :)

**Comment by ogizanagi** [comment]
Rebased to target 5.1

File: UPGRADE-5.1.md [hunk 2]
```diff
@@ -18,6 +18,12 @@ HttpFoundation
    `RedirectResponse::create()`, and `StreamedResponse::create()` methods (use
    `__construct()` instead)
 
+PropertyAccess
```
Comment (fabpot) [inline UPGRADE-5.1.md:21]: Should be moved to the 5.2 UPGRADE file

File: src/Symfony/Component/PropertyAccess/CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,12 @@
 CHANGELOG
 =========
 
+5.1.0
```
Comment (fabpot) [inline src/Symfony/Component/PropertyAccess/CHANGELOG.md:4]: 5.2.0

**Review by fabpot — CHANGES_REQUESTED** [review]
LGTM (not sure why I didn't review this PR before).
Can you rebase on current master, fix the minor issues so that I can merge?
Thank you.

**Comment by ogizanagi** [comment]
@fabpot No problem, done.
But I'd friendly ask @dunglas & @joelwurtz review, since the component has changed a lot since 5.1 and now relies on the PropertyInfo component. Thus, this new version of the PR also tweak the PropertyInfo `ReflectionExtractor`

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 4]
```diff
@@ -38,17 +38,23 @@
  */
 class PropertyAccessor implements PropertyAccessorInterface
 {
+    /** @var int Allow none of the magic methods */
+    public const DISALLOW_MAGIC_METHODS = ReflectionExtractor::DISALLOW_MAGIC_METHODS;
+    /** @var int Allow magic __get methods */
+    public const MAGIC_GET = ReflectionExtractor::ALLOW_MAGIC_GET;
```
Comment (ogizanagi) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php:44]: These are now aliases to the `ReflectionExtractor` constants, since it would be unexpected/not-discoverable for the user to use `ReflectionExtractor` consts to configure the accessor.

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php [hunk 5]
```diff
@@ -272,7 +291,7 @@ public function getReadInfo(string $class, string $property, array $context = []
     /**
      * {@inheritdoc}
      */
-    public function getWriteInfo(string $class, string $property, array $context = []): PropertyWriteInfo
+    public function getWriteInfo(string $class, string $property, array $context = []): ?PropertyWriteInfo
```
Comment (ogizanagi) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php:294]: This will go into another patch targeting 5.1 (some of the other changes in this class as well)

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php [hunk 6]
```diff
@@ -281,7 +300,16 @@ public function getWriteInfo(string $class, string $property, array $context = [
         }
 
         $allowGetterSetter = $context['enable_getter_setter_extraction'] ?? false;
-        $allowMagicCall = $context['enable_magic_call_extraction'] ?? false;
+        $magicMethods = $context['enable_magic_methods_extraction'] ?? $this->magicMethodsFlags;
+        $allowMagicCall = (bool) ($magicMethods & self::ALLOW_MAGIC_CALL);
+        $allowMagicSet = (bool) ($magicMethods & self::ALLOW_MAGIC_SET);
+
+        if (isset($context['enable_magic_call_extraction'])) {
+            trigger_deprecation('symfony/property-info', '5.2', 'Using the "enable_magic_call_extraction" context option in "%s()" is deprecated. Use "enable_magic_methods_extraction" instead.', __METHOD__);
```
Comment (ogizanagi) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php:308]: Other options are keeping this key and have single `enable_magic_get_extraction` / `enable_magic_set_extraction` as well, and either:
- have 3 new constructor arguments for the default
- keep having a single constructor arguments with flags, but keep having dedicated context keys.

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php [hunk 7]
```diff
@@ -73,14 +83,15 @@ class ReflectionExtractor implements PropertyListExtractorInterface, PropertyTyp
      * @param string[]|null $accessorPrefixes
      * @param string[]|null $arrayMutatorPrefixes
      */
-    public function __construct(array $mutatorPrefixes = null, array $accessorPrefixes = null, array $arrayMutatorPrefixes = null, bool $enableConstructorExtraction = true, int $accessFlags = self::ALLOW_PUBLIC, InflectorInterface $inflector = null)
+    public function __construct(array $mutatorPrefixes = null, array $accessorPrefixes = null, array $arrayMutatorPrefixes = null, bool $enableConstructorExtraction = true, int $accessFlags = self::ALLOW_PUBLIC, InflectorInterface $inflector = null, int $magicMethodsFlags = self::ALLOW_MAGIC_GET | self::ALLOW_MAGIC_SET)
```
Comment (ogizanagi) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php:86]: The `magicMethodsFlags` defaults are configured here for BC with 5.1, but before 5.1, magic get and set weren't supported, right?
So should it have been considered a BC break to enable them by default on 5.1?

**Comment by fabpot** [comment]
#37857 has been merged up to master, you can rebase now.

**Comment by ogizanagi** [comment]
fabbot failure is unrelated to this PR

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @ogizanagi.

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php (hunk 7 above)
Comment (xabbuh) [reply src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php:86]: @ogizanagi Yes, this looks like a BC break to me. I think we need to set the default value to `ReflectionExtractor::DISALLOW_MAGIC_METHODS`. We can consider deprecating this default value, but I am not convinced that's really necessary.
